### DISCUSSION TITLE
Missing environment variable causing monolog mailer exception error printed to browser in prod mode with debug false

### DISCUSSION THREAD

**Discussion by SherinBloemendaal** [symfony/symfony#57308] [Q&A]
Hi,

I've noticed that when using the monolog handler symfony_mailer, it prints the exception trace in the browser whenever an error occurs while sending an email. I've set APP_ENV=prod and APP_DEBUG=false. The stack trace includes a complete dump of the dotenv file, which is concerning to me. Currently, the mailer is misconfigured, but this issue could also arise if the mail server is down or experiencing temporary issues.

The total output of the error log is 8GB. Below, I've added a snippet for reference. As you can see, the print starts after the </html> tag is closed.

Is this expected behavior? I understand this might be an edge case since the error is thrown inside the error handler.

**UPDATE: I've noticed that the dump is caused when a environment variable is missing that is being used inside the `services.yaml`. Other errors (such as throwing an exception inside a controller) do not cause the dump. Still it seems weird that the stack trace including the entire dotenv is being dumped.**

php.ini:
```ini
log_errors=On
error_reporting=E_ALL & ~E_NOTICE & ~E_DEPRECATED & ~E_STRICT
display_errors=Off
```
php-fpm.conf:
```ini
[www]
...
catch_workers_output = yes
access.log = /proc/self/fd/2
user=www-data
group=www-data
```
 
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8" />
    <meta name="robots" content="noindex,nofollow,noarchive" />
    <title>An Error Occurred: Internal Server Error</title>
    <link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 128 128%22><text y=%221.2em%22 font-size=%2296%22>❌</text></svg>" />
    <style>body { background-color: #fff; color: #222; font: 16px/1.5 -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif; margin: 0; }
.container { margin: 30px; max-width: 600px; }
h1 { color: #dc3545; font-size: 24px; }
h2 { font-size: 18px; }</style>
</head>
<body>
<div class="container">
    <h1>Oops! An Error Occurred</h1>
    <h2>The server returned a "500 Internal Server Error".</h2>

    <p>
        Something is broken. Please let us know what you were doing when this error occurred.
        We will fix it as soon as possible. Sorry for any inconvenience caused.
    </p>
</div>
</body>
</html>object(Symfony\Component\Mailer\Exception\HttpTransportException)#885 (9) {
  ["message":protected]=>
  string(264) "Unable to send an email: The from address does not match a verified Sender Identity. Mail cannot be sent until this error is resolved. Visit https://sendgrid.com/docs/for-developers/sending-email/sender-identity/ to see the Sender Identity requirements (code 403)."
  ["string":"Exception":private]=>
  string(0) ""
  ["code":protected]=>
  int(0)
  ["file":protected]=>
  string(70) "/app/vendor/symfony/sendgrid-mailer/Transport/SendgridApiTransport.php"
  ["line":protected]=>
...
```

**Comment by MatTheCat** [comment]
Hello,

> I've set […] APP_DEBUG=false

Beware that [the string `false` is a truthy value](https://3v4l.org/jONmK#v8.3.7), so you may be running Symfony in debug mode.

**Reply by SherinBloemendaal** [reply]
Good one indeed.

I've executed a `dd()` on the `$context` variable that is being passed to the Kernel:
```php
  "APP_DEBUG" => "0"
```

It seems the symfony/runtime component uses `filter_var` on the APP_DEBUG variable:
https://github.com/symfony/runtime/blob/ea34522c447dd91a2b31cb330ee4540a56ba53f6/GenericRuntime.php#L68

I've replaced it anyways to prevent confusion, thank you!

**Reply by SherinBloemendaal** [reply]
It looks like it only happens on pages that do not exist.

I will try to create a reproduce. My current scenario is:
1. An environment variable is missing that's defined in the `services.yaml`
2. Using the monolog symfony_mailer with wrong credentials causing a HttpTransportException to be thrown
3. Calling a route that does not exist (resulting in a 404 Not found)

**Reply by MatTheCat** [reply]
Nice didn’t know about this `filter_var` :ok_hand: 

It may be that returning a 404 is just a way to trigger your monolog handler (if you didn’t [exclude it](https://symfony.com/doc/current/logging/monolog_exclude_http_codes.html)), so the issue would be that an exception thrown inside a monolog handler is dumped on the error page (which seems quite serious..!)

**Reply by SherinBloemendaal** [reply]
Indeed, i haven't excluded 404 errors (yet). The monolog handler is indeed called and the exception is thrown inside the mail handler
